packages/light-client-extension-helpers/src/smoldot/tasks.ts

```diff
@@ -0,0 +1,68 @@
+import { Client, AddChainError } from "."
+import { Effect, Fiber, Runtime, Schedule } from "effect"
+
+export const DEFAULT_SUPERVISE_REPEAT_SCHEDULE = Schedule.spaced("1 second")
+export const DEFAULT_SUPERVISE_RETRY_SCHEDULE = Schedule.spaced("1 second")
+
+export type SuperviseOptions = {
+  repeatSchedule: Schedule.Schedule<number>
+  retrySchedule: Schedule.Schedule<number>
+}
+
+/**
+ * Supervises a smoldot client by periodically invoking `addChain` with an
+ * empty chain spec.
+ *
+ * If `addChain` fails with anything other than a `AddChainError`, the client
+ * will be restarted.
+ *
+ * @param client - The smoldot client instance to be supervised.
+ * @param options - Options for customizing the supervision behavior.
+ * @param options.repeatSchedule - The frequency at which to invoke `addChain`.
+ * Defaults to {@link DEFAULT_SUPERVISE_REPEAT_SCHEDULE}.
+ * @param options.retrySchedule - The frequency at which to attempt restarting
+ * the client if needed. Defaults to {@link DEFAULT_SUPERVISE_RETRY_SCHEDULE}.
+ */
+export const supervise = async (
+  client: Client,
+  { repeatSchedule, retrySchedule }: SuperviseOptions = {
+    repeatSchedule: DEFAULT_SUPERVISE_REPEAT_SCHEDULE,
+    retrySchedule: DEFAULT_SUPERVISE_RETRY_SCHEDULE,
+  },
+) =>
+  Effect.gen(function* () {
+    const runtime = yield* Effect.runtime<never>()
+    const runPromise = Runtime.runPromise(runtime)
+
+    const daemon = yield* Effect.tryPromise(() =>
+      client
+        .addChain({ chainSpec: "" })
+        .then(() => {})
+        .catch((err) => {
+          if (err instanceof AddChainError) {
+            return
+          }
+
+          throw err
+        }),
+    ).pipe(
+      Effect.repeat({
+        schedule: repeatSchedule,
+      }),
+      Effect.tapError(() => Effect.tryPromise(() => client.restart())),
+      Effect.tapError(Effect.logError),
+      Effect.retry({
+        schedule: retrySchedule,
+      }),
+      Effect.forkDaemon,
+    )
+
+    const stop = () =>
+      Fiber.interrupt(daemon)
+        .pipe(Effect.andThen(() => Effect.void))
+        .pipe(runPromise)
+
+    return {
+      stop,
+    }
+  }).pipe(Effect.runPromise)
```

Comment:
When does something actually get the `stop` function back? ie does `.pipe(Effect.runPromise)` just run the gen thing forever?

It feels like you could kick off the daemon and return the stop method synchronously (esp because the only user of this function is not async anyways), but I don't understand this `Effect` stuff well enough to know what it's actually doing!

Reply:
it would be on `const { stop } = await supervise(client)`, but you're right that it can be sychronous. updated it to use `Effect.runSync` instead. 

We don't actually use the stop method but I think its just good to have.